filters/diag/logbody.go

```diff
@@ -0,0 +1,130 @@
+package diag
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/zalando/skipper/filters"
+	"github.com/zalando/skipper/filters/flowid"
+)
+
+type logBody struct {
+	limit    int
+	request  bool
+	response bool
+}
+
+// NewLogBody creates a filter specification for the 'logBody()' filter.
+func NewLogBody() filters.Spec { return logBody{} }
+
+// Name returns the logBody filtern name.
+func (logBody) Name() string {
+	return filters.LogBodyName
+}
+
+func (logBody) CreateFilter(args []interface{}) (filters.Filter, error) {
+	var (
+		request  = false
+		response = false
+	)
+
+	if len(args) != 2 {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	opt, ok := args[0].(string)
+	if !ok {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+	switch opt {
+	case "response":
+		response = true
+	case "request":
+		request = true
+	default:
+		return nil, fmt.Errorf("failed to match %q: %w", opt, filters.ErrInvalidFilterParameters)
+	}
+
+	limit, ok := args[1].(float64)
+	if !ok || float64(int(limit)) != limit {
+		return nil, fmt.Errorf("failed to convert to int: %w", filters.ErrInvalidFilterParameters)
+	}
+
+	return &logBody{
+		limit:    int(limit),
+		request:  request,
+		response: response,
+	}, nil
+}
+
+func (lb *logBody) Request(ctx filters.FilterContext) {
+	if !lb.request {
+		return
+	}
+
+	req := ctx.Request()
+	if req.Body != nil {
+		req.Body = newLogBodyStream(
+			lb.limit,
+			func(chunk []byte) {
+				ctx.Logger().Infof(
+					`logBody("request") %s: %s`,
+					req.Header.Get(flowid.HeaderName),
+					chunk)
+			},
+			req.Body,
+		)
+	}
+}
+
+func (lb *logBody) Response(ctx filters.FilterContext) {
+	if !lb.response {
+		return
+	}
+
+	rsp := ctx.Response()
+	if rsp.Body != nil {
+		rsp.Body = newLogBodyStream(
+			lb.limit,
+			func(chunk []byte) {
+				ctx.Logger().Infof(
+					`logBody("response") %s: %s`,
+					ctx.Request().Header.Get(flowid.HeaderName),
+					chunk)
+			},
+			rsp.Body,
+		)
+	}
+}
+
+type logBodyStream struct {
+	left  int
+	f     func([]byte)
+	input io.ReadCloser
+}
+
+func newLogBodyStream(left int, f func([]byte), rc io.ReadCloser) io.ReadCloser {
+	return &logBodyStream{
+		left:  left,
+		f:     f,
+		input: rc,
+	}
+}
+
+func (lb *logBodyStream) Read(p []byte) (n int, err error) {
+	if lb.left <= 0 {
+		return lb.input.Read(p)
+	}
+
+	n, err = lb.input.Read(p)
+	if n > 0 {
+		lb.f(p[:min(n, lb.left)])
+	}
+	lb.left -= n
```

Comment:
```go
n, err = lb.input.Read(p)
if n > 0 && lb.left > 0 {
    m := n
    if m > lb.left { // maybe we shouldn't bump go to 1.22 until its released
        m = lb.left
    }
    lb.f(p[:m])
    lb.left -= m
}
```

Reply:
I have go.mod set to 1.21 and min() seems to exist, so I left it as it was

Reply:
yeah, it was added in 1.21 https://pkg.go.dev/builtin#min